src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+      //     indexof(x,"A",0)!=-1 ^
+      //     indexof( x, "B", indexof( x, "A", 0 ) + 1 + 3 ) != -1 ^
+      //     indexof( x, "B", indexof( x, "A", 0 ) + 1 + 3 )+1+2 <= len(x)
+      return returnElim(atom, res, "concat-with-gaps");
+    }
+  }
+
+  if (!options::regExpElimAgg())
+  {
+    return Node::null();
+  }
+  // only aggressive rewrites below here
+
+  // if the first or last child is constant string, split
+  Node sStartIndex = d_zero;
+  Node sLength = lenx;
+  std::vector<Node> sConstraints;
+  std::vector<Node> rexpElimChildren;
+  for (unsigned r = 0; r < 2; r++)
+  {
+    unsigned index = r == 0 ? 0 : nchildren - 1;
+    Node c = children[index];
+    if (c.getKind() == STRING_TO_REGEXP)
+    {
+      Node s = c[0];
+      Node lens = nm->mkNode(STRING_LENGTH, s);
+      Node sss = r == 0 ? d_zero : nm->mkNode(MINUS, lenx, lens);
+      Node ss = nm->mkNode(STRING_SUBSTR, x, sss, lens);
+      sConstraints.push_back(ss.eqNode(s));
+      if (r == 0)
+      {
+        sStartIndex = lens;
+      }
+      sLength = nm->mkNode(MINUS, sLength, lens);
+    }
+    if (r == 1 && !sConstraints.empty())
+    {
+      // add the middle children
+      for (unsigned i = 1; i < (nchildren - 1); i++)
+      {
+        rexpElimChildren.push_back(children[i]);
+      }
+    }
+    if (c.getKind() != STRING_TO_REGEXP)
+    {
+      rexpElimChildren.push_back(c);
+    }
+  }
+  if (!sConstraints.empty())
+  {
+    Node ss = nm->mkNode(STRING_SUBSTR, x, sStartIndex, sLength);
+    Assert(!rexpElimChildren.empty());
+    Node regElim =
+        TheoryStringsRewriter::mkConcat(REGEXP_CONCAT, rexpElimChildren);
+    sConstraints.push_back(nm->mkNode(STRING_IN_REGEXP, ss, regElim));
+    Node ret = nm->mkNode(AND, sConstraints);
+    // e.g.
+    // x in re.++( "A", R ) ---> substr(x,0,1)="A" ^ substr(x,1,len(x)-1) in R
+    return returnElim(atom, ret, "concat-splice");
+  }
+  Assert(nchildren > 1);
+  for (unsigned i = 0; i < nchildren; i++)
+  {
+    if (children[i].getKind() == STRING_TO_REGEXP)
+    {
+      Node s = children[i][0];
+      Node lens = nm->mkNode(STRING_LENGTH, s);
+      // there exists an index in this string such that the substring is this
+      Node k;
+      std::vector<Node> echildren;
+      if (i == 0)
+      {
+        k = d_zero;
+      }
+      else if (i + 1 == nchildren)
+      {
+        k = nm->mkNode(MINUS, lenx, lens);
+      }
+      else
+      {
+        k = nm->mkBoundVar(nm->integerType());
+        Node bound =
+            nm->mkNode(AND,
+                       nm->mkNode(LEQ, d_zero, k),
+                       nm->mkNode(LT, k, nm->mkNode(MINUS, lenx, lens)));
+        echildren.push_back(bound);
+      }
+      Node substrEq = nm->mkNode(STRING_SUBSTR, x, k, lens).eqNode(s);
+      echildren.push_back(substrEq);
+      if (i > 0)
+      {
+        std::vector<Node> rprefix;
+        rprefix.insert(rprefix.end(), children.begin(), children.begin() + i);
```

Comment:
Nit: the previous statements can be merged to `std::vector<Node> rprefix(children.begin(), children.begin() + i);`